rs/rosetta-api/icp_ledger/ledger/tests/tests.rs

```diff
@@ -190,6 +191,80 @@ fn test_minting_account() {
     ic_icrc1_ledger_sm_tests::test_minting_account(ledger_wasm(), encode_init_args)
 }
 
+#[test]
+fn test_anonymous_transfers() {
+    const INITIAL_BALANCE: u64 = 10_000_000;
+    const TRANSFER_AMOUNT: u64 = 1_000_000;
+    let p1 = PrincipalId::new_user_test_id(1);
+    println!("p1: {:?}", p1);
+    let anon = PrincipalId::new_anonymous();
+    println!("anon: {:?}", anon);
+    let (env, canister_id) = setup(
+        ledger_wasm(),
+        encode_init_args,
+        vec![(Account::from(p1.0), INITIAL_BALANCE)],
+    );
+
+    assert_eq!(INITIAL_BALANCE, total_supply(&env, canister_id));
+    assert_eq!(INITIAL_BALANCE, balance_of(&env, canister_id, p1.0));
+    assert_eq!(0u64, balance_of(&env, canister_id, anon.0));
+
+    // Transfer to the account of the anonymous principal succeeds
+    transfer(&env, canister_id, p1.0, anon.0, TRANSFER_AMOUNT).expect("transfer failed");
+
+    // Transfer from the account of the anonymous principal using `icrc1_transfer` fails
+    let transfer_arg = TransferArg {
+        from_subaccount: None,
+        to: Account::from(p1.0),
+        fee: None,
+        created_at_time: None,
+        amount: Nat::from(TRANSFER_AMOUNT),
+        memo: None,
+    };
+    let encoded_transfer_result = env
+        .execute_ingress_as(
+            anon,
+            canister_id,
+            "icrc1_transfer",
+            Encode!(&transfer_arg).unwrap(),
+        )
+        .expect("failed to transfer funds")
+        .bytes();
+    let string_from_bytes_result = String::from_utf8(encoded_transfer_result.clone());
+    assert_eq!(
+        string_from_bytes_result,
+        Ok("Anonymous principal cannot hold tokens on the ledger.".to_string())
+    );
+
+    // Transfer from the account of the anonymous principal using the ICP-specific `transfer` fails
+    let transfer_args = icp_ledger::TransferArgs {
+        memo: icp_ledger::Memo(0u64),
+        amount: Tokens::from_e8s(TRANSFER_AMOUNT),
+        fee: Tokens::from_e8s(FEE),
+        from_subaccount: None,
+        to: AccountIdentifier::new(p1, None).to_address(),
+        created_at_time: None,
+    };
+    let response = env.execute_ingress_as(
+        anon,
+        canister_id,
+        "transfer",
+        Encode!(&transfer_args).unwrap(),
+    );
+    assert!(response.is_err());
+    if let Err(err) = response {
+        assert_eq!(err.code(), ErrorCode::CanisterCalledTrap);
+        assert!(err.description().contains("Canister called `ic0.trap` with message: Panicked at 'Sending from 2vxsx-fae is not allowed'"));
+    }
```

Comment:
Ideally, this `response` should be a `WasmResult`, [in this case containing an `icp_ledger::TransferError`](https://sourcegraph.com/github.com/dfinity/ic@490fbd87f2c97f9156ce1c266c6a665216f2bbfe/-/blob/rs/rosetta-api/icp_ledger/ledger/src/main.rs?L193), but since the ledger traps, we instead get a `UserError`.